src/muz/spacer/spacer_global_generalizer.cpp

```diff
@@ -125,39 +127,64 @@ void lemma_global_generalizer::operator()(lemma_ref &lemma) {
-static void to_real(expr_ref &fml) {
-    ast_manager &m = fml.get_manager();
-    arith_util arith(m);
-    array_util array(m);
-    if (arith.is_numeral(fml) || arith.is_to_real(fml)) return;
-    if (is_uninterp_const(fml) && arith.is_int(fml)) {
-        fml = arith.mk_to_real(fml);
-        return;
+struct to_real_rw_cfg : public default_rewriter_cfg {
+    ast_manager &m;
+    arith_util m_arith;
+    array_util m_array;
+    to_real_rw_cfg(ast_manager &m) : m(m), m_arith(m), m_array(m) {}
+    bool reduce_var(var *v, expr_ref &result, proof_ref &result_proof) {
+        if (m_arith.is_int(v))
+            result = m_arith.mk_to_real(v);
+        else
+            result = v;
+        return true;
     }
-    if (arith.is_arith_expr(fml)) {
-        app *fml_app = to_app(fml);
-        unsigned N = fml_app->get_num_args();
-        expr_ref_vector nw_args(m);
-        expr_ref chld(m);
-        for (unsigned i = 0; i < N; i++) {
-            chld = fml_app->get_arg(i);
-            to_real(chld);
-            nw_args.push_back(chld);
+
+    br_status reduce_app(func_decl *f, unsigned num, expr *const *args,
+                         expr_ref &result, proof_ref &result_pr) {
+        app *fml = nullptr;
+        if (num > 0)
+            // this call to mk_app creates a function whose domain has the same
+            // sort as args
+            fml = m.mk_app(f->get_family_id(), f->get_decl_kind(), num, args);
+        else {
+            // create const using same f
+            fml = m.mk_const(f);
+            if (is_uninterp_const(fml) && m_arith.is_int(fml)) {
+                result = m_arith.mk_to_real(fml);
+                return BR_DONE;
+            }
+            return BR_FAILED;
         }
-        fml = m.mk_app(fml_app->get_family_id(), fml_app->get_decl_kind(),
-                       nw_args.size(), nw_args.c_ptr());
-    }
-    if (array.is_select(fml)) {
-        app *fml_app = to_app(fml);
-        expr_ref_vector nw_args(m);
-        expr_ref ind(m);
-        ind = arith.mk_to_real(fml_app->get_arg(1));
-        nw_args.push_back(fml_app->get_arg(0));
-        nw_args.push_back(arith.mk_to_int(ind));
-        fml = array.mk_select(nw_args.size(), nw_args.c_ptr());
+
+        if (m_arith.is_to_real(fml) && m_arith.is_to_real(args[0])) {
+            result = args[0];
+            return BR_DONE;
+        }
+
+        if (m_array.is_select(fml)) {
+            expr_ref_vector new_args(m);
+            new_args.push_back(args[0]);
+            expr_ref ind(m);
+            for (unsigned i = 1; i < num; i++) {
+                if (!(m_arith.is_int_real(args[i]))) {
+                    TRACE("global",
+                          tout << "Found array with non arith index value\n";);
+                    NOT_IMPLEMENTED_YET();
+                    return BR_FAILED;
+                }
+                if (m_arith.is_real(args[i]))
+                    ind = m_arith.mk_to_int(args[i]);
```

Comment:
the use of `ind` creates an extra ref and dec of reference count. Put directly into reference counting buffer